pennylane/transforms/resolve_dynamic_wires.py

```diff
@@ -0,0 +1,184 @@
+from pennylane.measurements import measure
+from pennylane.tape import QuantumScript, QuantumScriptBatch
+from pennylane.typing import PostprocessingFn, Result, ResultBatch
+
+from .core import transform
+
+
+class _WireManager:
+
+    def __init__(self, zeroed=(), dirty=(), min_integer=None):
+        self._zeroed = list(zeroed)
+        self._dirty = list(dirty)
+        self._loaned = {}  # wire to final register type
+        self.min_integer = min_integer
+
+    def get_wire(self, require_zeros):
+        """Retrieve a concrete wire label from available registers."""
+        if not self._zeroed and not self._dirty:
+            if self.min_integer is None:
+                raise ValueError("no wires left to allocate.")
+            self._zeroed.append(self.min_integer)
+            self.min_integer += 1
+        if require_zeros:
+            if self._zeroed:
+                w = self._zeroed.pop()
+                self._loaned[w] = "zeroed"
+                return w, []
+            w = self._dirty.pop()
+            self._loaned[w] = "dirty"
+            m = measure(w, reset=True)
+            return w, m.measurements
+
+        if self._dirty:
+            w = self._dirty.pop()
+            self._loaned[w] = "dirty"
+            return w, []
+        w = self._zeroed.pop()
+        self._loaned[w] = "zeroed"
+        return w, []
+
+    def return_wire(self, wire, reset_to_original=False):
+        """Return a wire label back to be re-used."""
+        reg_type = self._loaned.pop(wire)
+        if reg_type == "zeroed" and reset_to_original:
+            self._zeroed.append(wire)
+        else:
+            self._dirty.append(wire)
+
+
+def null_postprocessing(results: ResultBatch) -> Result:
+    """An empty postprocessing function returned by resolve_dynamic_wires"""
+    return results[0]
+
+
+@transform
+def resolve_dynamic_wires(
+    tape: QuantumScript,
+    zeroed: Sequence[Hashable] = (),
+    dirty: Sequence[Hashable] = (),
+    min_integer: Optional[int] = None,
+) -> tuple[QuantumScriptBatch, PostprocessingFn]:
+    """Map dynamic wires to concrete values determined by the provided ``zeroed`` and ``dirty`` registers.
+
+    Args:
+        tape (QuantumScript): A circuit that may contain dynamic wire allocations and deallocations
+        zeroed (Sequence[Hashable]): a register of wires known to be the zero state
+        dirty (Sequence[Hashable]): a register of wires with any state
+
+    Returns:
+        tuple[QuantumScript], Callable[[ResultBatch], Result]: A batch of tapes and a postprocessing function
+
+    .. code-block:: python
+
+        @qml.qnode(qml.device('default.qubit'))
+        def circuit(, require_zeros=True):
+            with qml.allocation.safe_allocate(1, require_zeros=require_zeros) as wires:
+                qml.X(wires)
+            with qml.allocation.safe_allocate(1, require_zeros=require_zeros) as wires:
+                qml.Y(wires)
+            return qml.state()
```

Comment:
This example gives me a bit of a headache in terms of the return type :sweat_smile: 
Is this intentional?

The fact that the return type _shape_ will depend on the provided wires to `resolve_dynamic_wires` sounds like we have some technical difficulties coming our way :sweat_smile:

Reply:
I think a measurement is completely fine to have, but maybe one with definite wires, even the dynamic ones?